### ISSUE TITLE
Problema 5

### ISSUE THREAD

**Issue by josepdecid** [AsLogd/Problemes-Ampliaci-Algorismia#10]
Del problema 5 com enteneu que l'amplada i l'alçada dels rectangles és en binari? No veig en que pot ajudar a part d'indicar que són de tamanys enters.

**Comment by ghost** [comment]
Segurament si donessin la mida en una altre sistema la complexitat canviaria. (ex: hi havia un problema de classe que si l'entrada era en unari crec que passava a ser polinòmic).

**Comment by gonsp** [comment]
Pregunta importante: Los tiles que hay en T, se pueden repetir para rellenar el rectángulo? O solo se puede usar una vez cada tile que hay?

**Comment by ghost** [comment]
Desde mi punto de vista diría que son sin repetición (aunque puede haber 2 piezas iguales) y se permite rotaciones. Pero no hay información en el enunciado.

**Comment by josepdecid** [comment]
Entiendo que no se pueden repetir, lo que tampoco se es si se tiene que cubrir R con todos los tiles, o con un subconjunto de T  y que sobren algunos (creo que la segunda).

**Comment by gonsp** [comment]
Que no se puedan repetir lo haría más fácil hacer la demostración, pero no estoy seguro de eso. Fijaos que se habla de que T es un set. Por definición de set no podría haber 2 piezas iguales por ejemplo. Le mando un mail para asegurarnos.

**Comment by josepdecid** [comment]
Y lo de rotaciones, tiene sentido a parte de rotar 90º para poner en horizontal y vertical? Siendo rectángulos y que tienen que estar alineados con el rectángulo grande, no tiene sentido poderlos meter en diagonal.

**Comment by gonsp** [comment]
Osea que si T es un set, y queremos reducir 2-partition a este problema, no valdría ya que en 2-partition podrías tener 2 números iguales, pero en la entrada de este problema solo habría uno

**Comment by gonsp** [comment]
Solución:
2-Partition <= RECTANGLE TILING:
Dado un set S={x_1, x_2, ..., x_n}, y sea N = \sum x_i, creamos tiles de tamaño (1, x_1), (1, x_2), ..., (1, x_n) y un rectángulo de tamaño (2, N/2)

**Comment by gonsp** [comment]
Y ya está jajajaja al final resulta ser súper sencillo. Aunque esto es considerando que no hay rotaciones, porque si se permiten rotaciones y hay un número x_i = 2, entonces habría un tile que lo podrías poner vertical porque te cabe. Se soluciona fácilmente con una idea que tengo pero vamos a esperar a que carme responda diciendo si se permiten rotaciones y si puedes repetir tiles

**Comment by josepdecid** [comment]
Oh nice! Estaba pensando hacer algo similar unitario con la mochila y no acababa de ver como forzar el cover. Muy buena la idea del 2-Partition :raised_hands:

**Comment by adrian729** [comment]
Con hacer que R sea una sola fila de tamaño N/2 no sirve ya? Si puedes llenar el rectangulo la suma de los tiles que sobren seran exactamente N/2 tambien, y te evitas el problema que comentabas de tiles de tamaño 2 (lo de las rotaciones).
(Eso teniendo en cuenta el problema de Rectangle Tiling como que se tiene que llenar R y pueden sobrar Tiles en T)

**Comment by gonsp** [comment]
Mmm vale si, creo que sí que sirve eso

**Comment by ghost** [comment]
Se puede hacer lo que dice el @adrian729, pero sería necesario comprobar antes si N es par o no.
Si N es impar -> FALSE
Si N es par -> hacemos su transformación.
Así se evita el problema de rotaciones.

**Comment by Dexutar** [comment]
yes

**Comment by josepdecid** [comment]
Haría falta demostrar también que RECTANGLE TILING es NP no?

**Comment by gonsp** [comment]
Sep, pero bueno eso es trivial. Yo ya doy por solucionado este problema

**Comment by josepdecid** [comment]
Yep, vendría a ser algo así creo:

Dada una solución con las posiciones (esquina superior izquierda) de los tiles, existe un verificador en tiempo polinómico que dice si es válido el tiling `O(W*H)`
- Crear matriz booleana ocupación WxH a false
- Iterar por todas las tiles posicionadas y marcar a true las casillas que ocupa.
    - Si visitamos una posición ya ocupada, es que hay overlapping -> Solución inválida.
- Al acabar de iterar:
    - Ninguna casilla sin ocupar -> Solución válida.
    - Alguna casilla sin ocupar -> Solución inválida.

**Comment by ghost** [comment]
Dado los rectángulos usados, su posición izquierda superior y si está rotado o no se puede comprobar en P.

**Comment by JosepRivaille** [comment]
Escribo el pdf bien y lo subo en un rato si

**Comment by josepdecid** [comment]
Escribo el pdf y lo subo en un rato, así ya lo cerramos.

**Comment by gonsp** [comment]
Estoy pensando en qué pasa con que la entrada esté en binario. Si la entrada está en binario, y el tamaño de R se representa con (n, n) bits, necesitarás una matriz de tamaño (2^n)^2...

**Comment by gonsp** [comment]
Okay, definitivamente esto importa. Demostrar que es NP no es tan trivial porque no puedes hacer eso, ya que el testimonio se tiene que computar en tiempo polinómico. Si la entrada fuera dada en unario sí que sirve, pero en binario no. Esto se soluciona comprobando las superposiciones comparando todos los rectangulos y sus posiciones entre si

**Comment by ghost** [comment]
Cómo se comprueba que se rellena el tablero entero?
Si el tablero es de nxm bits acabará siendo 2^(n+m) posiciones a comprobar, no?

**Comment by arnaubadia** [comment]
Correcto, el tablero seria de 2^(height_tablero+width_tablero), comprobarlo no seria polinómico respecto a la entrada en binario.

Pero no hace falta expresar la solución como un tablero, puedes expresarla dando para cada tile T que pones en el la region R (el tablero) las coordenadas de inicio (top-left corner) y su tamaño (h x w). Entonces la solución tendría tamaño polinómico respecto la entrada.

Para que la solución sea válida hay que cumplir estas tres cosas:
1. Que todas las tiles de la solución estén dentro de la región.
2. Que la suma de areas de todas las tiles de la solución equivalga al area de la región.
3. Que ninguna tile se solape con otra (comprobar por cada par) .

Todas ellas se pueden hacer en tiempo polinómico.

Ya que el tamaño de la solución y el tiempo de verificación son polinómicos, el problema pertenece a NP.
(decidme si veis algún error)

**Comment by Dexutar** [comment]
yo lo veo bien

**Comment by ghost** [comment]
> Que la suma de areas de todas las tiles de la solución equivalga al area de la región.

Esta parte es la clave! Lo veo todo perfecto.

**Comment by DavidMoranPomes** [comment]
Pero, no haría falta demostrar que podemos llenar el tablero con un número polinómico de tiles? Porque si necesitamos un numero exponencial de tiles tendremos un testigo de tiempo exponencial también :(

**Comment by ghost** [comment]
Las tiles las coges de la entrada y no se pueden repetir, por lo tanto como mucho usas todas las que te dan.

**Comment by DavidMoranPomes** [comment]
No puedes repetir entonces? Ha contestado el mail Carme?

**Comment by DavidMoranPomes** [comment]
Us deixo en aquest [pdf](https://github.com/AsLogd/Problemes-Ampliaci-Algorismia/files/2098697/Examen_5.pdf) la meva versió d'aquest problema, formalitzant una mica tot el que hem comentat.
